### ISSUE TITLE
[web] flutter run -d chrome --debug does not boot with specific Record code

### ISSUE THREAD

**Issue by S-YOU** [flutter/flutter#129482]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Specific code that cause stuck (no window appear) flutter run in debug mode (release mode run fine)

-v mode show, program stopped at this line
> [   +4 ms] <- compile org-dartlang-app:/web_entrypoint.dart

code that cause this eror.
```dart
// ...
author: json['author'] != null ? ((json) => (name: json['name']))(json['author']) as BlogPageBlogEntriesAuthorRecord : null,
// ...
```

I don't know which part is wrong, ternary or closure or Record code or type or variable name.

Here is full function, it parse json into Record.
```dart
BlogPageRecord parseBlogPage(Map<String, dynamic> json) {
  return (
    blog: ((json) => (
          blogId: json['blogId'],
          title: json['title'],
          entries: json['entries']
              .map<BlogPageBlogEntriesRecord>((json) => (
                    authorId: json['authorId'],
                    body: json['body'],
                    entryId: json['entryId'],
                    title: json['title'],
                    author: json['author'] != null ? ((json) => (name: json['name']))(json['author']) as BlogPageBlogEntriesAuthorRecord : null,
                    // author: null,
                  ) as BlogPageBlogEntriesRecord)
              .toList(),
        ))(json['blog']) as BlogPageBlogRecord,
  );
}
```
If I change to author: null, it boots, so problem must be there.

typedefs are here for reference
```dart
typedef BlogPageRecord = ({
  BlogPageBlogRecord blog,
});
typedef BlogPageBlogRecord = ({
  String blogId,
  String title,
  List<BlogPageBlogEntriesRecord> entries,
});
typedef BlogPageBlogEntriesRecord = ({
  String authorId,
  String body,
  String entryId,
  String title,
  BlogPageBlogEntriesAuthorRecord? author,
});
typedef BlogPageBlogEntriesAuthorRecord = ({
  String name,
});
```

### Expected results

Browser window open and application is shown.

### Actual results

flutter run stuck and application does not start (no browser window open)

### Flutter Doctor output

Doctor output (I am not using iOS or android for now, only web(chrome) on Linux)

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.10.5, on Ubuntu 22.04.2 LTS 5.19.0-45-generic, locale C.UTF-8)
[✗] Android toolchain - develop for Android devices
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
[✓] Chrome - develop for the web
[✗] Linux toolchain - develop for Linux desktop
    ✗ GTK 3.0 development libraries are required for Linux development.
      They are likely available from your distribution (e.g.: apt install libgtk-3-dev)
[!] Android Studio (not installed)
[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
[✓] VS Code (version 1.79.2)
[✓] Connected device (2 available)
[✓] Network resources

! Doctor found issues in 3 categories.
```

**Comment by darshankawar** [comment]
Thanks for the report @S-YOU 
Can you provide us complete runnable code sample that we can directly copy paste and run to verify the reported behavior ?
Also please elaborate on actual vs expected behavior upon running the code.

**Comment by S-YOU** [comment]
Just copy paste the code in post to end of lib/main.dart and run with `flutter run -d chrome --debug
`. it stop working for me  @darshankawar 

Here is combined code for your reference.

```dart
typedef BlogPageRecord = ({
  BlogPageBlogRecord blog,
});
typedef BlogPageBlogRecord = ({
  String blogId,
  String title,
  List<BlogPageBlogEntriesRecord> entries,
});
typedef BlogPageBlogEntriesRecord = ({
  String authorId,
  String body,
  String entryId,
  String title,
  BlogPageBlogEntriesAuthorRecord? author,
});
typedef BlogPageBlogEntriesAuthorRecord = ({
  String name,
});

BlogPageRecord parseBlogPage(Map<String, dynamic> json) {
  return (
    blog: ((json) => (
          blogId: json['blogId'],
          title: json['title'],
          entries: json['entries']
              .map<BlogPageBlogEntriesRecord>((json) => (
                    authorId: json['authorId'],
                    body: json['body'],
                    entryId: json['entryId'],
                    title: json['title'],
                    author: json['author'] != null ? ((json) => (name: json['name']))(json['author']) as BlogPageBlogEntriesAuthorRecord : null,
                    // author: null,
                  ) as BlogPageBlogEntriesRecord)
              .toList(),
        ))(json['blog']) as BlogPageBlogRecord,
  );
}
```

flutter analyze say no issue found, btw.

**Comment by S-YOU** [comment]
Seems like you can even repro it in current Dartpad online, Compiler does not go back at all. @darshankawar 

Select [River Pod](https://dartpad.dev/?id=ef06ab3ce0b822e6cc5db0575248e6e2) example and add the codes in previous post at the end and click Run.

![image](https://github.com/flutter/flutter/assets/1001175/ca7c72b9-268a-47e4-aa89-73a7c795e70a)

**Comment by darshankawar** [comment]
Thanks for the update. I was able to replicate the reported behavior wherein, the tool gets stuck at:

```
flutter run -d chrome --debug
Launching lib/main.dart on Chrome in debug mode...
Waiting for connection from debug service on Chrome...
```

Even without passing `--debug` flag, the same behavior occurs. 

The same scenario runs properly on mobile wherein it compiles and builds the app properly.

Note that, we need to be on `sdk: ">=3.0.0 <4.0.0"` version to be able to use the [code sample provided](https://github.com/flutter/flutter/issues/129482#issuecomment-1607338026) to replicate the issue.

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.10.5, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.10.5 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (2 days ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.12.0-11.0.pre.32, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.12.0-11.0.pre.32 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7dbdacff46 (3 hours ago), 2023-06-26 22:05:23 -0400
    • Engine revision f320b8c36f
    • Dart version 3.1.0 (build 3.1.0-246.0.dev)
    • DevTools version 2.24.0
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by christopherfujino** [comment]
> The same scenario runs properly on mobile wherein it compiles and builds the app properly.

Sounds like this might be a DDC bug with records cc @annagrin 

@S-YOU have you tried to compile this in release mode for web?

**Comment by jason-simmons** [comment]
This app works in profile/release mode when compiled with dart2js.

But DDC spins forever while compiling the app if it includes the `author:` line

@sigmundch

**Comment by S-YOU** [comment]
> @S-YOU have you tried to compile this in release mode for web (which uses a different compiler)?

Yes, release mode run fine. 
(Only normal development with release mode is kind of inconvenience, so It was the reason I raised the issue.)
Didn't know it uses different compiler, that for the information. @christopherfujino

**Comment by christopherfujino** [comment]
> > @S-YOU have you tried to compile this in release mode for web (which uses a different compiler)?
> 
> Yes, release mode run fine. (Only normal development with release mode is kind of inconvenience, so It was the reason I raised the issue.) Didn't know it uses different compiler, that for the information. @christopherfujino

Yes, I'm glad you filed this. I was not suggesting you use release mode instead, this is just to help us narrow down the root cause.

**Comment by sigmundch** [comment]
Thanks so much for the repro!

cc @nshahan

It seems that the various casts added to the code are tripping the compiler. In particular, I was able to reduce it further to a small program that uses one cast:
```dart
main() {
  Object value = 'x';
  (a: 'a', b: value) as ({String a, String b});
}
```

(statically, the value is a `({String a, Object b})` record)

Strangely small tweaks elide the issue. For example, positional records don't trigger this, and swapping the values as `(a: value, b: 'b')` also escapes this problem.

As for the original program, the following suggestion is a bit ugly, but it appears you may be able to workaround the issue by adding a few casts in subexpressions, instead of adding a cast for the entire record. For example, this code no longer spins:

```dart
typedef BlogPageRecord = ({
  BlogPageBlogRecord blog,
});
typedef BlogPageBlogRecord = ({
  String blogId,
  String title,
  List<BlogPageBlogEntriesRecord> entries,
});
typedef BlogPageBlogEntriesRecord = ({
  String authorId,
  String body,
  String entryId,
  String title,
  BlogPageBlogEntriesAuthorRecord? author,
});
typedef BlogPageBlogEntriesAuthorRecord = ({
  String name,
});

BlogPageRecord parseBlogPage(Map<String, dynamic> json) {
  return (
    blog: ((json) => (
          blogId: json['blogId'] as String,
          title: json['title'] as String,
          entries: parseEntries(json['entries']),
        ))(json['blog']),
  );
}

List<BlogPageBlogEntriesRecord> parseEntries(List<dynamic> json) =>
    json.map<BlogPageBlogEntriesRecord>((e) => parseEntry(e)).toList();

BlogPageBlogEntriesRecord parseEntry(Map<String, dynamic> json) => (
      authorId: json['authorId'] as String,
      body: json['body'] as String,
      entryId: json['entryId'] as String,
      title: json['title'] as String,
      author: parseAuthor(json),
    );

BlogPageBlogEntriesAuthorRecord? parseAuthor(Map<String, dynamic> json) {
  return json['author'] != null
      ? ((json) => (name: json['name'] as String))(json['author'])
      : null;
}
```

**Comment by S-YOU** [comment]
Thank you very much for providing the workaround and pinpointing the issue, sigmundch. Appreciated for your assistance!

**Comment by nshahan** [comment]
This issue was fixed in the Dart SDK 3.0.6 which was part of the Flutter 3.10.6 release.

**Comment by S-YOU** [comment]
Thank you and team for the fix and release. 🎉
